### DISCUSSION TITLE
null-ls.health检测报错，如何解决？

### DISCUSSION THREAD

**Discussion by Jasonlib** [nshen/learn-neovim-lua#46] [General]
![image](https://user-images.githubusercontent.com/14918439/173480766-0c2f7272-97b1-4ce9-b376-d0dbb47f8e8f.png)

**Comment by nshen** [comment]
这是不同语言对应的格式化工具，根据需要安装对应的就好了，你现在有rustfmt就应该可以格式化rust代码了。

比如 stylelua 是格式化 lua 代码的，也可以去装一下 https://github.com/JohnnyMorganz/StyLua

**Comment by Jasonlib** [comment]
> 
每个格式化工具都需要手动安装吗？能不能配置自动安装呢？ [https://github.com/jose-elias-alvarez/null-ls.nvim/tree/main/lua/null-ls/builtins/formatting](url) 目录中的这些lua文件，是否和/null-ls.nvim插件配置有所关联呢？就是说，能否通过这些lua文件，来配置实现自动安装格式化工具呢？
![image](https://user-images.githubusercontent.com/14918439/173529322-bdcc7ca1-ff1c-4965-98ac-7b6eb26762b9.png)

**Comment by nshen** [comment]
应该还没有自动安装功能。那些lua文件只知道怎么调用对应的工具，需要手动安装后，在source里引用对应的lua文件。

```
 sources = {
        require("null-ls").builtins.formatting.stylua,
        require("null-ls").builtins.diagnostics.eslint,
        require("null-ls").builtins.completion.spell,
    },
```

**Comment by Jasonlib** [comment]
> 应该还没有自动安装功能。那些lua文件只知道怎么调用对应的工具，需要手动安装后，在source里引用对应的lua文件。
> 
> ```
>  sources = {
>         require("null-ls").builtins.formatting.stylua,
>         require("null-ls").builtins.diagnostics.eslint,
>         require("null-ls").builtins.completion.spell,
>     },
> ```

这个配置语句是要写在.config/nvim/lua/lspnull-ls.lua文件中吗？
还有prettier和eslint这两个格式化工具安装比较特殊，在.config/nvim/lua/lspnull-ls.lua中的如下配置语句，指示的node_modules文件夹应该放在什么位置？我试着将已经下载配置好的prettier和eslint格式化工具所在的node_modules文件夹，复制到config/nvim/lua/lspnull-ls.lua文件所在的同级目录：config/nvim/lua/ ，但是打开nvim checkhealth，仍然会报错。
但是如果通过软链接的方式，将prettier和eslint放到/usr/bin/全局目录，报错就会消失，但这么做应该会有问题吧？
![image](https://user-images.githubusercontent.com/14918439/173726506-c0d27404-e622-4290-bdf6-d0b8ff85a83b.png)
![image](https://user-images.githubusercontent.com/14918439/173727413-e6ae0c7e-7dcc-41d5-a654-a479a2074caf.png)

**Comment by nshen** [comment]
并不是所有前端都用 eslint 和prettier，所以 可以**不**全局安装，checkhealth 应该只会检测是否全局安装，所以报错也应该没关系。
设置 prefer_local 优先查找本地项目里是否有，没有再fallback到全局。

https://github.com/jose-elias-alvarez/null-ls.nvim/blob/main/doc/BUILTIN_CONFIG.md#using-local-executables
